README.md
# Scikit-qfit
Package that fits gradient orthogonal Q polynomials to freeform shapes


CHANGES.rst
=========
Changelog
=========

Version 0.1.5
=============
- Fix factorial, factorial2 scipy import

Version 0.1.4
=============
- Adds spectrum to topography functionality
- Updated documentation and minor fixes

Version 0.1.3
=============
- Fixed error when data has invalids

Version 0.1.2
=============
- Fixed error in interpolator call
- Added documentation to readthedocs

Version 0.1.1
=============

- Corrections and addition to PyPi documentation
- Added example file

Version 0.1
===========

- Initial public pre-alpha release



tests/conftest.py
# -*- coding: utf-8 -*-
"""
    Dummy conftest.py for skqfit.

    If you don't know what this is for, just leave it empty.
    Read more about conftest.py under:
    https://pytest.org/latest/plugins.html
"""

# import pytest


tests/test_synthmap.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
Test the Q spectrum processing using a synthesised image as specified
in Section 5 of the "Fitting freeform shapes with orthogonal bases" document.
"""

from __future__ import print_function, absolute_import, division

__author__ = "npdata"
__copyright__ = "npdata"
__license__ = "mit"

import math
import time
import numpy as np
from numpy.polynomial.polynomial import polyval2d
from skqfit.qspectre import QSpectrum

def eval_synthmap(as_map=False, inverse=True):
    def sag_fn(rhov, thetav):
        x = rhov*np.cos(thetav)
        y = rhov*np.sin(thetav)
        return sag_xy(x, y)

    def sag_xy(x, y):
        cx = -1/452.62713
        cy = -1/443.43539
        kx = ky = 0.0
        x2 = x*x
        y2 = y*y
        z = (cx*x2 + cy*y2)/(1 + np.sqrt(1-((1+kx)*cx*cx)*x2 - ((1+ky)*cy*cy)*y2))
        return z + polyval2d(x, y, sag_xy.coeff)

    sag_xy.rmax = 174.2
    sag_xy.curv = -1/478.12597
    sag_xy.coeff = np.zeros((11,11), dtype=np.float)
    sag_xy.coeff[0,:]  = [0, 0, 8.446692E-05,	-1.773111E-08,	2.103339E-10,	-4.450410E-14,	1.204820E-15,	-3.751270E-18,	1.243271E-20,	-1.671689E-23,	2.740074E-26]
    sag_xy.coeff[2,:9] = [6.086975E-05,	-9.657166E-08,	3.881972E-10,	-5.340721E-13,	1.962740E-15,	-3.972902E-18,	2.276418E-20,	-6.515923E-23,	1.259617E-25]
    sag_xy.coeff[4,:7] = [1.345443E-10,	-4.424293E-13,	1.672236E-15,	-4.286471E-18,	1.613314E-20,	-4.548523E-23,	9.938038E-26]
    sag_xy.coeff[6,:5] = [3.310262E-16,	-1.749391E-18,	7.515349E-21,	-2.305324E-23,	1.939290E-26]
    sag_xy.coeff[8,:3] = [1.020537E-21,	-6.739667E-24,	-3.800397E-27]
    sag_xy.coeff[10,0] = 1.653756E-28

    def build_map(pts, slice=False):
        x = np.linspace(-1.02*sag_xy.rmax, 1.02*sag_xy.rmax, pts)
        if slice:
            y = np.linspace(0.0, 0.0, 1)
        else:
            y = np.linspace(-1.02*sag_xy.rmax, 1.02*sag_xy.rmax, pts)
        xv, yv = np.meshgrid(x, y, indexing='ij')
        z = sag_xy(xv, yv)
        return x, y, z.reshape((x.size, y.size))

    def test_xy_gradient(zmap, dfdx, dfdy, x, y):
        grad = np.gradient(zmap)
        gx = grad[0] / (x[1] - x[0])
        dy = grad[1] / (y[1] - y[0])

        err_dx = dfdx - gx
        gx_err = np.nanmax(err_dx[1:-1,1:-1]) - np.nanmin(err_dx[1:-1,1:-1])
        err_dy = dfdy - dy
        gy_err = np.nanmax(err_dy[1:-1,1:-1]) - np.nanmin(err_dy[1:-1,1:-1])

        #display_map(err_dx[1:-1,1:-1])
        #display_map(err_dy[1:-1,1:-1])
        return max(gx_err, gy_err)

    exp_ispec = np.array([[70531,   225291, 25895,  199399, 3583,   2651,   1886,   339,    55, 41, 5],
                          [43,	    223995,	11377,	198,	2604,	801,	46,	    37,	    5,	0,	0],
                          [82,	    12916,	3592,	994,	158,	10,	    5,	    0,	    0,	0,	0],
                          [10,	    1568,	256,	10,	    2,	    0,	    0,	    0,	    0,	0,	0],
                          [1,	    20,	    3,	    1,	    0,	    0,	    0,	    0,	    0,	0,	0]], dtype=np.int)


    bfs_c  = sag_xy.curv
    points = 501
    if False:
        m_max = 200
        n_max = 200
    else:
        m_max = 10
        n_max = 9
    qfit = QSpectrum(m_max, n_max)

    if as_map:
        x, y, zmap = build_map(points)
        qfit.data_map(x, y, zmap, centre=(0.,0.), radius=sag_xy.rmax)
        #display_map(zmap)
    else:
        qfit.set_sag_fn(sag_fn, sag_xy.rmax, bfs_c)

    start = time.time()
    a_nm, b_nm = qfit.q_fit(m_max, n_max)
    print('fit done, time %.3fs' % (time.time() - start))
    qspec = np.sqrt(np.square(a_nm) + np.square(b_nm))
    #disp_qspec(qspec)

    ispec = np.round(1e6*qspec).astype(int)
    idiff = ispec[:5,:11] - exp_ispec
    errors = np.count_nonzero(idiff)
    inv_err, grad_err = 0.0, 0.0

    if inverse:
        if not as_map:
            x, y, zmap = build_map(points)
        start = time.time()
        zinv, dfdx, dfdy = qfit.build_map(x, y, radius=sag_xy.rmax, centre=(0.0,0.0), a_nm=a_nm, b_nm=b_nm, interpolated=True, inc_deriv=True)
        print('inverse done, time %.3fs' % (time.time() - start))
        grad_err = test_xy_gradient(zinv, dfdx, dfdy, x, y)
        cond = zinv != 0.0
        diff = np.extract(cond, zmap - zinv)
        inv_err = max(math.fabs(np.nanmax(diff)), math.fabs(np.nanmin(diff)))

    assert errors == 0 and inv_err < 1.0e-7 and grad_err < 1.0e-5

def test_as_map():
    eval_synthmap(as_map=True, inverse=True)
    
def test_as_sagfn():
    eval_synthmap(as_map=False, inverse=True)
    
if __name__ == "__main__":
    test_as_sagfn()
    test_as_map()



skqfit/asmjacp.py
"""
References:

G W Forbes, "Characterizing the shape of freeform optics", Opt. Express 20(3), 2483-2499 (2012)

"""


from __future__ import print_function, absolute_import, division

import numpy as np

class AsymJacobiP(object):
    """
    Generate asymmetric Jacobi like polynominals needed for the freeform fit as
    defined in the reference document [A.1]

    
    Warning
    -------
    The Jacobi P polynominals can generate large values that lead to a 
    double overflow for large m and n values. It is tested to (1500, 1500) for (m,n).
    Exceeding these values may lead to overflow events.

    Note
    ----
    The scipi.special.jacobi function is has a 1.5:1 performance advantage over the
    current implementation but it doesn't support the normalization and can lead 
    to an overflow condition for n or m over 500. 
    """

    def __init__(self, nmax):
        self.scan_n = None
        self.a_mn = np.zeros(nmax, dtype=np.float)
        self.b_mn = np.zeros_like(self.a_mn)
        self.c_mn = np.zeros_like(self.a_mn)
        self.abc_mn = None
        self.m = None
        self.nmax = nmax

    def build_recursion(self, m):
        """
        Build the recurssion coefficients and saves them as a sequence of tuples.
        These are the coefficients to build the m type polynominals up to order n.
        """
        start = 1 if m > 1 else 3
        self.scan_n = range(start, self.nmax)
        self.a_mn.fill(0.0)
        self.b_mn.fill(0.0)
        self.c_mn.fill(0.0)
        for n in self.scan_n:
            d_mn = (4.0*n*n-1)*(m+n-2)*(m+2*n-3)
            self.a_mn[n] = (2.0*n-1)*(m+2*n-2)*(4*n*(m+n-2)+(m-3)*(2*m-1))/d_mn
            self.b_mn[n] = -2.0*(2*n-1)*(m+2*n-1)*(m+2*n-2)*(m+2*n-3)/d_mn
            self.c_mn[n] = n*(2.0*n-3)*(m+2*n-1)*(2*m+2*n-3)/d_mn
        self.abc_mn = list(zip(self.scan_n, self.a_mn[start:], self.b_mn[start:], self.c_mn[start:]))
        self.m = m

    def jvec_x(self, jvec, x):
        """
        Builds the sequence of jacobi polynomials for the value x
        based on [A.2-5]
        """
        m = self.m
        if m == 1:
            jvec[0] = 0.5
            jvec[1] = 1.0 - x/2
            jvec[2] = (3.0 + x*(-12 + 8*x))/6 
            jvec[3] = (5.0 + x*(-60 + (120 - 64*x)*x))/10.0
            vn_ = jvec[2]
            vn  = jvec[3]
        else:
            jvec[0] = 0.5
            jvec[1] = (m - 0.5) + (1 - m)*x
            vn_ = jvec[0]
            vn  = jvec[1]
        for (n,a,b,c) in self.abc_mn:
            vn_, vn = vn, (a + b*x)*vn - c*vn_
            jvec[n+1] = vn
        return jvec

    def jmat_x(self, jmat, xv):
        """
        Builds the asymmetric Jacobi P polynomial as defined in [2] A.1 for 
        all a vector of x values.
        """
        m = self.m
        if m == 1:
            jmat[0,:].fill(0.5)
            jmat[1,:] = 1.0 - xv/2
            jmat[2,:] = (3.0 + xv*(-12 + 8*xv))/6 
            jmat[3,:] = (5.0 + xv*(-60 + (120 - 64*xv)*xv))/10.0
            vn_ = jmat[2,:]
            vn  = jmat[3,:]
        else:
            jmat[0,:].fill(0.5)
            jmat[1,:] = (m - 0.5) + (1 - m)*xv
            vn_ = jmat[0,:]
            vn  = jmat[1,:]
        for (n,a,b,c) in self.abc_mn:
            vn_, vn = vn, (a + b*xv)*vn - c*vn_
            jmat[n+1,:] = vn
        return jmat

    def jmat_u_x(self, jmat, uv, xv):
        """
        Builds the asymmetric Jacobi P polynomial as defined in [2] A.1 for all of the
        x values with the scaling factor of u**m which extends the range of usable (m,n)
        values before an overflow condition occurs.
        """
        m = self.m
        if m == 1:
            jmat[0,:].fill(0.5)
            jmat[1,:] = (1.0 - xv/2)
            jmat[2,:] = ((3.0 + xv*(-12 + 8*xv))/6) 
            jmat[3,:] = ((5.0 + xv*(-60 + (120 - 64*xv)*xv))/10.0)
            vn_ = jmat[2,:]
            vn  = jmat[3,:]
        else:
            jmat[0,:].fill(0.5)
            jmat[1,:] = uv*((m - 0.5) + (1 - m)*xv)
            vn_ = jmat[0,:]
            vn  = jmat[1,:]
        for (n,a,b,c) in self.abc_mn:
            n_ = n+1
            if n_ < m:
                vn_, vn = vn, uv*((a + b*xv)*vn - uv*c*vn_)
            elif n_ == m:
                vn_, vn = vn, (a + b*xv)*vn - uv*c*vn_                  
            else:
                vn_, vn = vn, (a + b*xv)*vn - c*vn_                  
            jmat[n_,:] = vn

        # Complete the uv**m scaling of the data 
        n = max(0, m - 1 - (jmat.shape[0]-1))
        upm = np.power(uv, n)
        for i in range(min(jmat.shape[0]-1, m-1), -1, -1):
            jmat[i,:] *= upm
            upm *= uv
        return jmat

    def jvec_u_x(self, jvec, u, x):
        """
        Builds the asymmetric Jacobi P polynomial for a single x as defined in [2] A.1 for all of the
        x values with the scaling factor of u**m which extends the range of usable (m,n) values.
        """
        m = self.m
        if m == 1:
            jvec[0] = 0.5
            jvec[1] = (1.0 - x/2)
            jvec[2] = ((3.0 + x*(-12 + 8*x))/6) 
            jvec[3] = ((5.0 + x*(-60 + (120 - 64*x)*x))/10.0)
            vn_ = jvec[2]
            vn  = jvec[3]
        else:
            jvec[0] = 0.5
            jvec[1] = u*((m - 0.5) + (1 - m)*x)
            vn_ = jvec[0]
            vn  = jvec[1]
        for (n,a,b,c) in self.abc_mn:
            n_ = n+1
            if n_ < m:
                vn_, vn = vn, u*((a + b*x)*vn - u*c*vn_)
            elif n_ == m:
                vn_, vn = vn, (a + b*x)*vn - u*c*vn_                  
            else:
                vn_, vn = vn, (a + b*x)*vn - c*vn_      
                           
            jvec[n_] = vn

        # Complete the u**m scaling of the data 
        n = max(0, m - 1 - (jvec.size-1))
        upm = u**n
        for i in range(min(jvec.size-1, m-1), -1, -1):
            jvec[i] *= upm
            upm *= u
        return jvec


skqfit/_version.py

# This file helps to compute a version number in source trees obtained from
# git-archive tarball (such as those provided by githubs download-from-tag
# feature). Distribution tarballs (built by setup.py sdist) and build
# directories (produced by setup.py build) will contain a much shorter file
# that just contains the computed version number.

# This file is released into the public domain. Generated by
# versioneer-0.15 (https://github.com/warner/python-versioneer)

import errno
import os
import re
import subprocess
import sys


def get_keywords():
    # these strings will be replaced by git during git-archive.
    # setup.py/versioneer.py will grep for the variable names, so they must
    # each be defined on a line of their own. _version.py will just call
    # get_keywords().
    git_refnames = "$Format:%d$"
    git_full = "$Format:%H$"
    keywords = {"refnames": git_refnames, "full": git_full}
    return keywords


class VersioneerConfig:
    pass


def get_config():
    # these strings are filled in when 'setup.py versioneer' creates
    # _version.py
    cfg = VersioneerConfig()
    cfg.VCS = "git"
    cfg.style = "pep440"
    cfg.tag_prefix = ""
    cfg.parentdir_prefix = "skqfit"
    cfg.versionfile_source = "skqfit/_version.py"
    cfg.verbose = False
    return cfg


class NotThisMethod(Exception):
    pass


LONG_VERSION_PY = {}
HANDLERS = {}


def register_vcs_handler(vcs, method):  # decorator
    def decorate(f):
        if vcs not in HANDLERS:
            HANDLERS[vcs] = {}
        HANDLERS[vcs][method] = f
        return f
    return decorate


def run_command(commands, args, cwd=None, verbose=False, hide_stderr=False):
    assert isinstance(commands, list)
    p = None
    for c in commands:
        try:
            dispcmd = str([c] + args)
            # remember shell=False, so use git.cmd on windows, not just git
            p = subprocess.Popen([c] + args, cwd=cwd, stdout=subprocess.PIPE,
                                 stderr=(subprocess.PIPE if hide_stderr
                                         else None))
            break
        except EnvironmentError:
            e = sys.exc_info()[1]
            if e.errno == errno.ENOENT:
                continue
            if verbose:
                print("unable to run %s" % dispcmd)
                print(e)
            return None
    else:
        if verbose:
            print("unable to find command, tried %s" % (commands,))
        return None
    stdout = p.communicate()[0].strip()
    if sys.version_info[0] >= 3:
        stdout = stdout.decode()
    if p.returncode != 0:
        if verbose:
            print("unable to run %s (error)" % dispcmd)
        return None
    return stdout


def versions_from_parentdir(parentdir_prefix, root, verbose):
    # Source tarballs conventionally unpack into a directory that includes
    # both the project name and a version string.
    dirname = os.path.basename(root)
    if not dirname.startswith(parentdir_prefix):
        if verbose:
            print("guessing rootdir is '%s', but '%s' doesn't start with "
                  "prefix '%s'" % (root, dirname, parentdir_prefix))
        raise NotThisMethod("rootdir doesn't start with parentdir_prefix")
    return {"version": dirname[len(parentdir_prefix):],
            "full-revisionid": None,
            "dirty": False, "error": None}


@register_vcs_handler("git", "get_keywords")
def git_get_keywords(versionfile_abs):
    # the code embedded in _version.py can just fetch the value of these
    # keywords. When used from setup.py, we don't want to import _version.py,
    # so we do it with a regexp instead. This function is not used from
    # _version.py.
    keywords = {}
    try:
        f = open(versionfile_abs, "r")
        for line in f.readlines():
            if line.strip().startswith("git_refnames ="):
                mo = re.search(r'=\s*"(.*)"', line)
                if mo:
                    keywords["refnames"] = mo.group(1)
            if line.strip().startswith("git_full ="):
                mo = re.search(r'=\s*"(.*)"', line)
                if mo:
                    keywords["full"] = mo.group(1)
        f.close()
    except EnvironmentError:
        pass
    return keywords


@register_vcs_handler("git", "keywords")
def git_versions_from_keywords(keywords, tag_prefix, verbose):
    if not keywords:
        raise NotThisMethod("no keywords at all, weird")
    refnames = keywords["refnames"].strip()
    if refnames.startswith("$Format"):
        if verbose:
            print("keywords are unexpanded, not using")
        raise NotThisMethod("unexpanded keywords, not a git-archive tarball")
    refs = set([r.strip() for r in refnames.strip("()").split(",")])
    # starting in git-1.8.3, tags are listed as "tag: foo-1.0" instead of
    # just "foo-1.0". If we see a "tag: " prefix, prefer those.
    TAG = "tag: "
    tags = set([r[len(TAG):] for r in refs if r.startswith(TAG)])
    if not tags:
        # Either we're using git < 1.8.3, or there really are no tags. We use
        # a heuristic: assume all version tags have a digit. The old git %d
        # expansion behaves like git log --decorate=short and strips out the
        # refs/heads/ and refs/tags/ prefixes that would let us distinguish
        # between branches and tags. By ignoring refnames without digits, we
        # filter out many common branch names like "release" and
        # "stabilization", as well as "HEAD" and "master".
        tags = set([r for r in refs if re.search(r'\d', r)])
        if verbose:
            print("discarding '%s', no digits" % ",".join(refs-tags))
    if verbose:
        print("likely tags: %s" % ",".join(sorted(tags)))
    for ref in sorted(tags):
        # sorting will prefer e.g. "2.0" over "2.0rc1"
        if ref.startswith(tag_prefix):
            r = ref[len(tag_prefix):]
            if verbose:
                print("picking %s" % r)
            return {"version": r,
                    "full-revisionid": keywords["full"].strip(),
                    "dirty": False, "error": None
                    }
    # no suitable tags, so version is "0+unknown", but full hex is still there
    if verbose:
        print("no suitable tags, using unknown + full revision id")
    return {"version": "0+unknown",
            "full-revisionid": keywords["full"].strip(),
            "dirty": False, "error": "no suitable tags"}


@register_vcs_handler("git", "pieces_from_vcs")
def git_pieces_from_vcs(tag_prefix, root, verbose, run_command=run_command):
    # this runs 'git' from the root of the source tree. This only gets called
    # if the git-archive 'subst' keywords were *not* expanded, and
    # _version.py hasn't already been rewritten with a short version string,
    # meaning we're inside a checked out source tree.

    if not os.path.exists(os.path.join(root, ".git")):
        if verbose:
            print("no .git in %s" % root)
        raise NotThisMethod("no .git directory")

    GITS = ["git"]
    if sys.platform == "win32":
        GITS = ["git.cmd", "git.exe"]
    # if there is a tag, this yields TAG-NUM-gHEX[-dirty]
    # if there are no tags, this yields HEX[-dirty] (no NUM)
    describe_out = run_command(GITS, ["describe", "--tags", "--dirty",
                                      "--always", "--long"],
                               cwd=root)
    # --long was added in git-1.5.5
    if describe_out is None:
        raise NotThisMethod("'git describe' failed")
    describe_out = describe_out.strip()
    full_out = run_command(GITS, ["rev-parse", "HEAD"], cwd=root)
    if full_out is None:
        raise NotThisMethod("'git rev-parse' failed")
    full_out = full_out.strip()

    pieces = {}
    pieces["long"] = full_out
    pieces["short"] = full_out[:7]  # maybe improved later
    pieces["error"] = None

    # parse describe_out. It will be like TAG-NUM-gHEX[-dirty] or HEX[-dirty]
    # TAG might have hyphens.
    git_describe = describe_out

    # look for -dirty suffix
    dirty = git_describe.endswith("-dirty")
    pieces["dirty"] = dirty
    if dirty:
        git_describe = git_describe[:git_describe.rindex("-dirty")]

    # now we have TAG-NUM-gHEX or HEX

    if "-" in git_describe:
        # TAG-NUM-gHEX
        mo = re.search(r'^(.+)-(\d+)-g([0-9a-f]+)$', git_describe)
        if not mo:
            # unparseable. Maybe git-describe is misbehaving?
            pieces["error"] = ("unable to parse git-describe output: '%s'"
                               % describe_out)
            return pieces

        # tag
        full_tag = mo.group(1)
        if not full_tag.startswith(tag_prefix):
            if verbose:
                fmt = "tag '%s' doesn't start with prefix '%s'"
                print(fmt % (full_tag, tag_prefix))
            pieces["error"] = ("tag '%s' doesn't start with prefix '%s'"
                               % (full_tag, tag_prefix))
            return pieces
        pieces["closest-tag"] = full_tag[len(tag_prefix):]

        # distance: number of commits since tag
        pieces["distance"] = int(mo.group(2))

        # commit: short hex revision ID
        pieces["short"] = mo.group(3)

    else:
        # HEX: no tags
        pieces["closest-tag"] = None
        count_out = run_command(GITS, ["rev-list", "HEAD", "--count"],
                                cwd=root)
        pieces["distance"] = int(count_out)  # total number of commits

    return pieces


def plus_or_dot(pieces):
    if "+" in pieces.get("closest-tag", ""):
        return "."
    return "+"


def render_pep440(pieces):
    # now build up version string, with post-release "local version
    # identifier". Our goal: TAG[+DISTANCE.gHEX[.dirty]] . Note that if you
    # get a tagged build and then dirty it, you'll get TAG+0.gHEX.dirty

    # exceptions:
    # 1: no tags. git_describe was just HEX. 0+untagged.DISTANCE.gHEX[.dirty]

    if pieces["closest-tag"]:
        rendered = pieces["closest-tag"]
        if pieces["distance"] or pieces["dirty"]:
            rendered += plus_or_dot(pieces)
            rendered += "%d.g%s" % (pieces["distance"], pieces["short"])
            if pieces["dirty"]:
                rendered += ".dirty"
    else:
        # exception #1
        rendered = "0+untagged.%d.g%s" % (pieces["distance"],
                                          pieces["short"])
        if pieces["dirty"]:
            rendered += ".dirty"
    return rendered


def render_pep440_pre(pieces):
    # TAG[.post.devDISTANCE] . No -dirty

    # exceptions:
    # 1: no tags. 0.post.devDISTANCE

    if pieces["closest-tag"]:
        rendered = pieces["closest-tag"]
        if pieces["distance"]:
            rendered += ".post.dev%d" % pieces["distance"]
    else:
        # exception #1
        rendered = "0.post.dev%d" % pieces["distance"]
    return rendered


def render_pep440_post(pieces):
    # TAG[.postDISTANCE[.dev0]+gHEX] . The ".dev0" means dirty. Note that
    # .dev0 sorts backwards (a dirty tree will appear "older" than the
    # corresponding clean one), but you shouldn't be releasing software with
    # -dirty anyways.

    # exceptions:
    # 1: no tags. 0.postDISTANCE[.dev0]

    if pieces["closest-tag"]:
        rendered = pieces["closest-tag"]
        if pieces["distance"] or pieces["dirty"]:
            rendered += ".post%d" % pieces["distance"]
            if pieces["dirty"]:
                rendered += ".dev0"
            rendered += plus_or_dot(pieces)
            rendered += "g%s" % pieces["short"]
    else:
        # exception #1
        rendered = "0.post%d" % pieces["distance"]
        if pieces["dirty"]:
            rendered += ".dev0"
        rendered += "+g%s" % pieces["short"]
    return rendered


def render_pep440_old(pieces):
    # TAG[.postDISTANCE[.dev0]] . The ".dev0" means dirty.

    # exceptions:
    # 1: no tags. 0.postDISTANCE[.dev0]

    if pieces["closest-tag"]:
        rendered = pieces["closest-tag"]
        if pieces["distance"] or pieces["dirty"]:
            rendered += ".post%d" % pieces["distance"]
            if pieces["dirty"]:
                rendered += ".dev0"
    else:
        # exception #1
        rendered = "0.post%d" % pieces["distance"]
        if pieces["dirty"]:
            rendered += ".dev0"
    return rendered


def render_git_describe(pieces):
    # TAG[-DISTANCE-gHEX][-dirty], like 'git describe --tags --dirty
    # --always'

    # exceptions:
    # 1: no tags. HEX[-dirty]  (note: no 'g' prefix)

    if pieces["closest-tag"]:
        rendered = pieces["closest-tag"]
        if pieces["distance"]:
            rendered += "-%d-g%s" % (pieces["distance"], pieces["short"])
    else:
        # exception #1
        rendered = pieces["short"]
    if pieces["dirty"]:
        rendered += "-dirty"
    return rendered


def render_git_describe_long(pieces):
    # TAG-DISTANCE-gHEX[-dirty], like 'git describe --tags --dirty
    # --always -long'. The distance/hash is unconditional.

    # exceptions:
    # 1: no tags. HEX[-dirty]  (note: no 'g' prefix)

    if pieces["closest-tag"]:
        rendered = pieces["closest-tag"]
        rendered += "-%d-g%s" % (pieces["distance"], pieces["short"])
    else:
        # exception #1
        rendered = pieces["short"]
    if pieces["dirty"]:
        rendered += "-dirty"
    return rendered


def render(pieces, style):
    if pieces["error"]:
        return {"version": "unknown",
                "full-revisionid": pieces.get("long"),
                "dirty": None,
                "error": pieces["error"]}

    if not style or style == "default":
        style = "pep440"  # the default

    if style == "pep440":
        rendered = render_pep440(pieces)
    elif style == "pep440-pre":
        rendered = render_pep440_pre(pieces)
    elif style == "pep440-post":
        rendered = render_pep440_post(pieces)
    elif style == "pep440-old":
        rendered = render_pep440_old(pieces)
    elif style == "git-describe":
        rendered = render_git_describe(pieces)
    elif style == "git-describe-long":
        rendered = render_git_describe_long(pieces)
    else:
        raise ValueError("unknown style '%s'" % style)

    return {"version": rendered, "full-revisionid": pieces["long"],
            "dirty": pieces["dirty"], "error": None}


def get_versions():
    # I am in _version.py, which lives at ROOT/VERSIONFILE_SOURCE. If we have
    # __file__, we can work backwards from there to the root. Some
    # py2exe/bbfreeze/non-CPython implementations don't do __file__, in which
    # case we can only use expanded keywords.

    cfg = get_config()
    verbose = cfg.verbose

    try:
        return git_versions_from_keywords(get_keywords(), cfg.tag_prefix,
                                          verbose)
    except NotThisMethod:
        pass

    try:
        root = os.path.realpath(__file__)
        # versionfile_source is the relative path from the top of the source
        # tree (where the .git directory might live) to this file. Invert
        # this to find the root from __file__.
        for i in cfg.versionfile_source.split('/'):
            root = os.path.dirname(root)
    except NameError:
        return {"version": "0+unknown", "full-revisionid": None,
                "dirty": None,
                "error": "unable to find root of source tree"}

    try:
        pieces = git_pieces_from_vcs(cfg.tag_prefix, root, verbose)
        return render(pieces, cfg.style)
    except NotThisMethod:
        pass

    try:
        if cfg.parentdir_prefix:
            return versions_from_parentdir(cfg.parentdir_prefix, root, verbose)
    except NotThisMethod:
        pass

    return {"version": "0+unknown", "full-revisionid": None,
            "dirty": None,
            "error": "unable to compute version"}


skqfit/skeleton.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
This is a skeleton file that can serve as a starting point for a Python
console script. To run this script uncomment the following line in the
entry_points section in setup.cfg:

    console_scripts =
        hello_world = skqfit.module:function

Then run `python setup.py install` which will install the command `hello_world`
inside your current environment.
Besides console scripts, the header (i.e. until _logger...) of this file can
also be used as template for Python modules.

Note: This skeleton file can be safely removed if not needed!
"""
from __future__ import division, print_function, absolute_import

import argparse
import sys
import logging

from skqfit import __version__

__author__ = "npdata"
__copyright__ = "npdata"
__license__ = "mit"

_logger = logging.getLogger(__name__)


def parse_args(args):
    """
    Parse command line parameters

    :param args: command line parameters as list of strings
    :return: command line parameters as :obj:`argparse.Namespace`
    """
    parser = argparse.ArgumentParser(
        description="Just a Hello World demonstration")
    parser.add_argument(
        '-v',
        '--version',
        action='version',
        version='Scikit-Qfit {ver}'.format(ver=__version__))
    return parser.parse_args(args)


def main(args):
    args = parse_args(args)
    print("Hello World!")
    _logger.info("Script ends here")


def run():
    logging.basicConfig(level=logging.INFO, stream=sys.stdout)
    main(sys.argv[1:])


if __name__ == "__main__":
    run()


example/qspectrum.py
"""
Processing example using a Mahr file format (txt).

The code loads a file, builds the q-spectrum and displays the image.

Note that invalids are replaced with NAN.

"""

import sys
import time
import getopt
import numpy as np
from matplotlib import pyplot as plt
from mpl_toolkits.axes_grid1 import make_axes_locatable

import skqfit.qspectre as qf

def load_mahr_file(filename):
    """
    loads a mahr file in txt format and returns the result as a zmap, x, y
    """

    need = ('START_X','STOP_X','START_Y','STOP_Y','INTERVAL_LIN', 'NO_TOT_POINTS')
    with open(filename,'r') as file:
        fpm={}
        line = file.readline()
        if line.strip() != "[PROFILE_HEADER]":
            raise Exception("Invalid File Format")

        # get the file parameters first
        while line.strip() != "[PROFILE_VALUES]":
            spl = line.split("=")
            if spl[0] in need:
                if len(spl) < 2:
                    raise Exception("Missing parameter: "+spl[0])
                else:
                    fpm[spl[0]] = float(spl[1])
            line = file.readline()
        if len(fpm) != len(need):
            raise Exception("Missing parameters")

        # now get the actual data
        rows = round((fpm['STOP_X']-fpm['START_X'])/fpm['INTERVAL_LIN']) + 1
        cols = round((fpm['STOP_Y']-fpm['START_Y'])/fpm['INTERVAL_LIN']) + 1
        pts =  round(float(fpm['NO_TOT_POINTS']))
        if rows * cols != pts:
            raise Exception("Unexpected number of points")

        zmap = np.zeros((rows, cols), dtype=np.float)
        x_0 = float(fpm['START_X'])
        y_0 = float(fpm['START_Y'])
        x_n = float(fpm['STOP_X'])
        y_n = float(fpm['STOP_Y'])
        spacing = float(fpm['INTERVAL_LIN'])
      
        def do_line(line):
            ll = line.split('=')
            if len(ll) == 2:
                vals = ll[1].split(' ')
                if len(vals) == 5 and int(vals[4]) == 0:
                    r = int(round((float(vals[0]) - x_0)/spacing))
                    c = int(round((float(vals[1]) - y_0)/spacing))
                    zmap[r,c] = float(vals[2])

        zmap.fill(np.nan)
        lines = file.readlines()
        for line in lines:
            do_line(line)

        x=np.linspace(x_0, x_n, rows)
        y=np.linspace(y_0, y_n, cols)

    return zmap, x, y

def disp_qspec(qmap, skip_rotnl=False, skip_interp=True, skip_low_order=False):
    """
    Display the Q-spectrum as a simple 2D map

    """
    fig, ax = plt.subplots()
    interp = 'none' if skip_interp else None      
    ax.xaxis.tick_top()
    ax.set_xlabel('Azimuthal (M)')    
    ax.xaxis.set_label_position('top')
    ax.set_ylabel('Radial (N)')
    srow = 0
    scol = 0
    if skip_rotnl:
        scol = 1
    if skip_low_order:
        srow = 1
    qm = qmap[srow:,scol:]
    peak_val = np.log10(np.max(qm))
    log_max = peak_val + 0.1
    log_min = log_max - 4.001
    offset = 10**(log_min-1)
    dmap = np.log10(qm + offset)
    im = plt.imshow(dmap, interpolation=interp)
    divider = make_axes_locatable(ax)
    cax2 = divider.append_axes("right", size="5%", pad=0.05)
    plt.colorbar(im, orientation='vertical', cax=cax2)
    plt.show()

def filter(fname, m_max=None, n_max=None):
    zmap, x, y = load_mahr_file(fname)
    qfit = qf.QSpectrum(m_max=m_max, n_max=n_max)
    qfit.data_map(x=x, y=y, zmap=zmap)
    a_nm, b_nm = qfit.q_fit(m_max, n_max)

    # Apply suitable filter in q-space
    # In this case just remove the low order azimuthal and radial
    mask = np.ones_like(a_nm)
    mask[:,0] = 0.0
    mask[0] = 0.0
    a_nm *= mask
    b_nm *= mask

    zinv = qfit.build_map(x, y, a_nm=a_nm, b_nm=b_nm, curv=0.0)

    im = plt.imshow(zinv)
    plt.colorbar(im, orientation='vertical')
    plt.show()

def residual(fname, m_max=None, n_max=None):
    zmap, x, y = load_mahr_file(fname)
    qfit = qf.QSpectrum(m_max=m_max, n_max=n_max)
    qfit.data_map(x=x, y=y, zmap=zmap)
    a_nm, b_nm = qfit.q_fit(m_max, n_max)

    zinv = qfit.build_map(x, y, a_nm=a_nm, b_nm=b_nm, extend=0.95)
    cond = zinv != 0.0
    diff = np.extract(cond, zmap - zinv)
    diff -= np.mean(diff)

    zmap.fill(np.nan)
    np.place(zmap, cond, diff)
    im = plt.imshow(zmap)
    plt.colorbar(im, orientation='vertical')
    plt.show()

def qspectrum(fname, m_max=None, n_max=None, show_rotnl=False):
    zmap, x, y = load_mahr_file(fname)
    start = time.time()
    qmap = qf.qspec(x, y, zmap, m_max=m_max, n_max=n_max)
    print('%s: spectrum done, time %.3fs' % (fname, time.time()-start))
    disp_qspec(qmap, skip_rotnl=not show_rotnl, skip_interp=True, skip_low_order=False)

if __name__ == "__main__":
    def usage():
        print('qspectrum.py --mmax=500 --nmax=500 --rotnl=False [qspec|resid|filter] filename')
    try:
        opts, args = getopt.getopt(sys.argv[1:],"",["mmax=", "nmax=", "rotnl="])
        process = args[0]
        fname = args[1]
    except (getopt.GetoptError, IndexError):
        usage()
        sys.exit(2)
    nmax = mmax = 500
    rotnl = False
    for o, a in opts:
        if o == "--mmax":
            mmax = int(a)
        elif o == "--nmax":
            nmax = int(a)
        elif o == "--rotnl":
            rotnl = (a == "True")
        else:
            usage()
            sys.exit()
    if process == "qspec":
        qspectrum(fname, m_max=mmax, n_max=nmax, show_rotnl=rotnl)
    elif process == "resid":
        residual(fname, m_max=mmax, n_max=nmax)
    elif process == "filter":
        filter(fname, m_max=mmax, n_max=nmax)
    else:
        usage()
        sys.exit()
